### PR TITLE
Draft for compiling NovusCore on macos

### PR DESCRIPTION
**Pull request by CarwynNelson**
This is a draft commit for building NovusCore on macos, working towards
building on linux. To keep things simple I am only building the
authserver and common.

So far the build is a bit janky, but it works and Nix has been able to
connect from his client to the authserver.

In order to generate the project files using cmake I ran the following commands:
```
mkdir build
cd build
cmake -DOPENSSL_ROOT_DIR=/usr/local/opt/openssl -DMYSQL_INCLUDE_DIR=/usr/local/mysql/include -DMYSQL_LIB_DIR=/usr/local/mysql/lib ..
```

It is my belief that `/usr/local/opt/openssl` is the default location for the openssl header files on a mac, but I'll need to check that on a blank mac to be certain. It just so happens that I have a mac that needs wiping so maybe I'll get around to that.

The mysql header files were downloaded as per the current setup instructions and they were located in my downloads folder. This did compile but oddly would not run, due to the following error:
```
dyld: Library not loaded: /usr/local/mysql/lib/libmysqlclient.18.dylib
  Referenced from: /Users/nelsoc55/code/NovusCore/build/./authserver
  Reason: image not found
[1]    61361 abort      ./authserver
```

The quick fix for that issue seems to be moving the mysql header files to `/usr/local/mysql`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -27,6 +27,7 @@ cmake_minimum_required(VERSION 3.12)
 set(CMAKE_DISABLE_SOURCE_CHANGES ON)
 set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
 set(CMAKE_CONFIGURATION_TYPES Debug Release RelWithDebInfo CACHE STRING "" FORCE)
+set(CMAKE_CXX_STANDARD 17)
```
Comment (CarwynNelson) [inline CMakeLists.txt]: For some reason, clang was not listening to the c++ standards set elsewhere. This was a quick and dirty way of telling clang to use c++17 for all of the projects.

I imagine we can find a cleaner and more proper way of doing this.

File: dep/libmpq/CMakeLists.txt [hunk 2]
```diff
@@ -24,9 +24,9 @@ file(GLOB sources libmpq/*.c libmpq/*.h)
 add_library(mpq STATIC ${sources})
 set_target_properties(mpq PROPERTIES LINKER_LANGUAGE CXX)
 
-if(WIN32)
+# if(WIN32)
```
Comment (CarwynNelson) [inline dep/libmpq/CMakeLists.txt]: It looks like this is also required on a mac.

Maybe we could have a `/mac` and `/linux` folders for the final change, or just use the same file for all of the platforms.

I guess this depends on if there is platform specific config.

File: dep/mysql/CMakeLists.txt [hunk 3]
```diff
@@ -32,5 +32,6 @@ else()
 endif()
 
 target_include_directories(mysql SYSTEM INTERFACE "${MYSQL_INCLUDE_DIR}")
-target_link_libraries(mysql INTERFACE "${MYSQL_LIB_DIR}/libmysql.lib")
-target_compile_definitions(mysql INTERFACE _ALLOW_ITERATOR_DEBUG_LEVEL_MISMATCH)
\ No newline at end of file
+# target_link_libraries(mysql INTERFACE "${MYSQL_LIB_DIR}/libmysql.lib")
+target_link_libraries(mysql INTERFACE "${MYSQL_LIB_DIR}/libmysqlclient.18.dylib")
```
Comment (CarwynNelson) [inline dep/mysql/CMakeLists.txt]: I guess we should wrap this in an `IF(MACOS)`. It would be nicer if we could tell cmake to locate this for us, but I'm not sure if that's possible?

File: projects/CMakeLists.txt [hunk 4]
```diff
@@ -22,8 +22,8 @@
 
 # Include Project Directories
 add_subdirectory(common)
-add_subdirectory(tools)
+# add_subdirectory(tools)
 add_subdirectory(authserver)
```
Comment (CarwynNelson) [inline projects/CMakeLists.txt:26]: As per the description, we are only building the authserver and common for this draft pull request, just so that we can make quick progress and figure out a plan for the whole project.

File: projects/authserver/ConnectionHandlers/ClientAuthConnectionHandler.h [hunk 5]
```diff
@@ -24,7 +24,7 @@
 #pragma once
 
 #include <Networking/TcpServer.h>
-#include "..\Connections\AuthConnection.h"
+#include "../Connections/AuthConnection.h"
```
Comment (CarwynNelson) [inline projects/authserver/ConnectionHandlers/ClientAuthConnectionHandler.h]: Clang seems to have an issue with `\` includes, and this seems to be where most of the errors came from.

It may be a bit pedantic but I feel that changing all of the `\` to `/` should be in their own commit (or even pull request) as it is somewhat unrelated to the cross-platform work.

File: projects/common/Database/DatabaseConnector.cpp [hunk 6]
```diff
@@ -83,7 +83,8 @@ void DatabaseConnector::Borrow(DATABASE_TYPE type, std::function<void(std::share
         _connectorPools[type].add(std::unique_ptr<DatabaseConnector>(newConnector));
     }
 
-    func(std::shared_ptr<DatabaseConnector>(_connectorPools[type].acquire()));
```
Comment (CarwynNelson) [inline projects/common/Database/DatabaseConnector.cpp:86]: This was a weird bug, but I get the impression that it points to a wider design problem with this class? At least that is what I was getting from the Discord conversation.

In any case, this appears to patch over it for this pull request.

File: projects/common/Math/Vector2.h [hunk 7]
```diff
@@ -34,22 +34,24 @@ class Vector2
...
-	Vector2(u32 in) { x = static_cast<f32>(in); y = static_cast<f32>(in); }
-	Vector2(i8 inX, i8 inY) { x = static_cast<f32>(inX); y = static_cast<f32>(inY); }
-	Vector2(i8 in) { x = static_cast<f32>(in); y = static_cast<f32>(in); }
-	Vector2(i16 inX, i16 inY) { x = static_cast<f32>(inX); y = static_cast<f32>(inY); }
-	Vector2(i16 in) { x = static_cast<f32>(in); y = static_cast<f32>(in); }
-	Vector2(i32 inX, i32 inY) { x = static_cast<f32>(inX); y = static_cast<f32>(inY); }
-	Vector2(i32 in) { x = static_cast<f32>(in); y = static_cast<f32>(in); }
-
-	Vector2(Vector2& in) { x = in.x; y = in.y; }
+	Vector2(float x, float y) : x(x), y(y) {}
```
Comment (CarwynNelson) [inline projects/common/Math/Vector2.h:37]: Clang seemed to get very confused by all of these different constructors and would refuse to compile. I've just commented out the ones that are not used for now and it solves the problem but this one might require further research.

File: projects/common/Math/Vector3.h [hunk 8]
```diff
@@ -34,22 +34,23 @@ class Vector3
...
-	Vector3(i8 inX, i8 inY, f32 inZ) { x = static_cast<f32>(inX); y = static_cast<f32>(inY); z = static_cast<f32>(inZ); }
-	Vector3(i8 in) { x = static_cast<f32>(in); y = static_cast<f32>(in); z = static_cast<f32>(in); }
-	Vector3(i16 inX, i16 inY, f32 inZ) { x = static_cast<f32>(inX); y = static_cast<f32>(inY); z = static_cast<f32>(inZ); }
-	Vector3(i16 in) { x = static_cast<f32>(in); y = static_cast<f32>(in); z = static_cast<f32>(in); }
-	Vector3(i32 inX, i32 inY, f32 inZ) { x = static_cast<f32>(inX); y = static_cast<f32>(inY); z = static_cast<f32>(inZ); }
-	Vector3(i32 in) { x = static_cast<f32>(in); y = static_cast<f32>(in); z = static_cast<f32>(in); }
-
-	Vector3(Vector3& in) { x = in.x; y = in.y; z = in.z; }
+	Vector3(float x, float y, float z) : x(x), y(y), z(z) {}
+	// Vector3(f32 inX, f32 inY, f32 inZ) { x = inX; y = inY; z = inZ; }
```
Comment (CarwynNelson) [inline projects/common/Math/Vector3.h:38]: Same as the comment in `Vector2.h`

File: projects/common/Utils/DebugHandler.h [hunk 9]
```diff
@@ -23,7 +23,10 @@
 */
 #pragma once
 
+#ifdef _WIN32
```
Comment (CarwynNelson) [inline projects/common/Utils/DebugHandler.h:26]: I think we need to figure out a way to write a more cross-platform DebugHandler. Maybe there are some libraries we can take advantage of to sort most of this out for us? spdlog maybe?

File: projects/common/Utils/DebugHandler.h [hunk 10]
```diff
@@ -37,82 +40,81 @@ enum PROGRAM_TYPE
...
+DebugHandler::PrintSuccess(message);
 
 class DebugHandler
 {
 public:
     static bool isInitialized;
     static void Initialize();
 
-    template <typename... Args>
-    inline static void Print(std::string message, Args... args)
```
Comment (CarwynNelson) [inline projects/common/Utils/DebugHandler.h:64]: Clang complained about this too, so I just removed it. As per the previous comment we might want to think about reworking this so that it is cross-platform and cross-compiler by default.

File: CMakeLists.txt (hunk 1 above)
Comment (NixAJ) [reply CMakeLists.txt]: This is now added by default, there is no reason for us to not have the entire project build as c++17

File: dep/libmpq/CMakeLists.txt (hunk 2 above)
Comment (NixAJ) [reply dep/libmpq/CMakeLists.txt]: This is fixed now

File: dep/mysql/CMakeLists.txt (hunk 3 above)
Comment (NixAJ) [reply dep/mysql/CMakeLists.txt]: I split this up into two ifs, we might need a different solution if .dylib isn't universal for unix

**Comment by MrDmitry** [comment]
Instead of `-DOPENSSL_ROOT_DIR=...` you could modify [NovusCore/dep/openssl/CMakeLists.txt](https://github.com/novuscore/NovusCore/blob/master/dep/openssl/CMakeLists.txt) to set default path both for `UNIX` and `APPLE`, not just `UNIX`. Also I think it should check if `OPENSSL_ROOT_DIR` was set by the user instead of silently overwriting it.

Also I'd suggest using [FindMySQL.cmake](https://github.com/facebook/hhvm/blob/master/CMake/FindMySQL.cmake) the same way openssl is accessed - via `find_package` instead of [relying on MYSQL_INCLUDE_DIR](https://github.com/novuscore/NovusCore/blob/master/dep/mysql/CMakeLists.txt)

File: projects/authserver/ConnectionHandlers/ClientAuthConnectionHandler.h (hunk 5 above)
Comment (NixAJ) [reply projects/authserver/ConnectionHandlers/ClientAuthConnectionHandler.h]: This was has been addressed

**Comment by CarwynNelson** [comment]
> Instead of `-DOPENSSL_ROOT_DIR=...` you could modify [NovusCore/dep/openssl/CMakeLists.txt](https://github.com/novuscore/NovusCore/blob/master/dep/openssl/CMakeLists.txt) to set default path both for `UNIX` and `APPLE`, not just `UNIX`. Also I think it should check if `OPENSSL_ROOT_DIR` was set by the user instead of silently overwriting it.

I am now able to run cmake without any parameters, but I assume that there are improvements that can be made to the existing cmake files to make them even more flexible.

File: projects/common/Utils/DebugHandler.h (hunk 9 above)
Comment (Pursche) [reply projects/common/Utils/DebugHandler.h:26]: We don't need a library for this, the functionality of the DebugHandler is really simple.
First we need to figure out variadric templates and how to pass them along to printf on Unix, this would be the "Args" part of the templates. If I followed the discord conversation correctly this works a bit differently in gcc/clang. Then we need to replace the Windows specific "SetConsoleTextAttribute" with whatever the Unix equivalent is within simple ifdefs.

File: projects/common/Utils/DebugHandler.h (hunk 10 above)
Comment (Pursche) [reply projects/common/Utils/DebugHandler.h:64]: Yes, I think the issue is that Clang doesn't like us not passing any args. So the whole "you can pass any amount of arguments here" becomes "you can pass any amount of arguments here, except 0" for some incredibly stupid reason. Maybe it's enough to copy it and change the copys template line to "template <>". I can't really look into this more detailed than this at the moment. :/

File: projects/common/Math/Vector2.h (hunk 7 above)
Comment (Pursche) [reply projects/common/Math/Vector2.h:37]: I will have to get my Clang running and look into fixing this, but I'm not sure why this would happen and I don't know if I have an opportunity to look at it until worst case Thursday.

File: projects/common/Database/DatabaseConnector.cpp (hunk 6 above)
Comment (Pursche) [reply projects/common/Database/DatabaseConnector.cpp:86]: This is just supposed to call the callback function "func" while passing a database connector into it. I am unsure which part of this fails, you should check if calling func with modified parameters to not pass anything into it is allowed, or maybe if doing the .aquire() outside of the parameters is allowed.

**Comment by Pursche** [comment]
This is looking good in general, I don't _think_ it would be particularly hard to get Clang to play nice with what we have written so far, I added comments with things you can try on the specific issues. Good work so far and I am sorry that I can't be more involved in this at the moment. :(

**Comment by CarwynNelson** [comment]
I am closing this as obsolete in favor of #12 .

There are still a number of things that have been raised out of this work that I will try to address as separate issues.

If it's okay with the project maintainers I will raise these as Github issues as these are easier to track for specific issues alongside pull requests.

I'll look to raise those in the next few days / week.

File: projects/common/Database/DatabaseConnector.cpp (hunk 6 above)
Comment (CarwynNelson) [reply projects/common/Database/DatabaseConnector.cpp:86]: I'm not quite certain what is meant by this, would you be able to explain this to me at some point?

Comment (NixAJ) [reply projects/common/Database/DatabaseConnector.cpp:86]: https://github.com/novuscore/NovusCore/blob/master/projects/common/Database/DatabaseConnector.cpp#L86

This is what Pursche meant
